src/android/java/io/jxcore/node/BtConnectorHelper.java

```diff
@@ -49,313 +51,281 @@ public BtConnectorHelper() {
+            throw new RuntimeException("BeginConnectPeer callback is NULL !!!!!!");
+        }
 
-            if (selectedDevice != null) {
-                if (mBTConnector != null) {
-                    BTConnector.TryConnectReturnValues retVal = mBTConnector.TryConnect(selectedDevice);
-                    if (retVal == BTConnector.TryConnectReturnValues.Connecting) {
-                        //all is ok, lets wait callbacks, and for them lets copy the callback here
-                        mConnectStatusCallback = connectStatusCallback;
-                    } else if (retVal == BTConnector.TryConnectReturnValues.NoSelectedDevice) {
-                        // we do check this already, thus we should not get this ever.
-                        tmpCallback.ConnectionStatusUpdate("Device Address for " + toPeerId + " not found from Discovered device list.",-1);
-                    } else if (retVal == BTConnector.TryConnectReturnValues.AlreadyAttemptingToConnect) {
-                        tmpCallback.ConnectionStatusUpdate("There is already one connection attempt progressing.",-1);
-                    } else if (retVal == BTConnector.TryConnectReturnValues.BTDeviceFetchFailed) {
-                        tmpCallback.ConnectionStatusUpdate("Bluetooth API failed to get Bluetooth device for the address : " + selectedDevice.peerAddress,-1);
-                    }
+        BtToRequestSocket tmpToReqSoc = mBtToRequestSocket;
+        if (tmpToReqSoc != null) {
+            connectStatusCallback.ConnectionStatusUpdate("Maximum peer connections reached, please try again after disconnecting a peer. Connected to " + tmpToReqSoc.GetPeerId(), -1);
+            return;
+        }
 
-                } else {
-                    tmpCallback.ConnectionStatusUpdate("Device conenctivity not started, please call StartBroadcasting before attempting to connnect",-1);
-                }
-            } else {
-                tmpCallback.ConnectionStatusUpdate("Device Address for " + toPeerId + " not found from Discovered device list.",-1);
+        ServiceItem selectedDevice = null;
+
+        Iterator<ServiceItem> iterator = lastAvailableList.iterator();
+        while (iterator.hasNext()) {
+            ServiceItem item = iterator.next();
+            if (item != null && item.peerId.contentEquals(toPeerId)) {
+                selectedDevice = item;
+                break;
             }
         }
+
+        if (selectedDevice == null) {
+            connectStatusCallback.ConnectionStatusUpdate("Device Address for " + toPeerId + " not found from Discovered device list.", -1);
+            return;
+        }
+
+        BTConnector tmpConn = mBTConnector;
+        if (tmpConn == null) {
+            connectStatusCallback.ConnectionStatusUpdate("Device connectivity not started, please call StartBroadcasting before attempting to connect", -1);
+            return;
+        }
+
+        BTConnector.TryConnectReturnValues retVal = tmpConn.TryConnect(selectedDevice);
```

Comment:
This whole section from 178 to 188 should be replaced with a switch statement. This is literally what they were invented for.

```Java
        switch (tmpConn.TryConnect(selectedDevice)) {
            case Connecting:
                mConnectStatusCallback = connectStatusCallback;
                break;

            etc.

            default: throw new RuntimeException("Unrecognized enum type!");
        }
```